server/src/main/java/com/cloud/api/ApiServer.java

```diff
@@ -456,7 +456,11 @@ public void handle(final HttpRequest request, final HttpResponse response, final
                         responseType = param.getValue();
                         continue;
                     }
-                    parameterMap.put(param.getName(), new String[]{param.getValue()});
+                    if(parameterMap.putIfAbsent(param.getName(), new String[]{param.getValue()}) != null) {
+                        String message = String.format("Query parameter '%s' has multiple values", param.getName());
+                        s_logger.error(message);
+                        throw new ServerApiException(ApiErrorCode.MALFORMED_PARAMETER_ERROR, message);
```

Comment:
this is a repeating bit (also happening in ApuServlet) maybe we can move (parts of) it to a common utility?